### PR TITLE
abstract multi-file clouder reader and apply to parquet

### PR DESCRIPTION
**Pull request by wbo4958**
This PR first abstracted the basic logic of multi-files cloud reader to be the common class that should be applied for different file formats, eg. Parquet, ORC, CSV, and so on.

Second, reformed the existing parquet multi-files cloud reader which extends the common class abstracted 

The initialization design doc can be found at https://docs.google.com/document/d/10sIOPMesnSNPvbb_SGnbE1C4gQv-G4ISorgYK8jQQns/edit#heading=h.ykdv684u0rnx

### CODE REVIEW

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
build

**Comment by tgravescs** [comment]
please add description and pointer to issue if one exists

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  }
+
+  protected def fileSystemBytesRead(): Long = {
+    FileSystem.getAllStatistics.asScala.map(_.getThreadStatistics.getBytesRead).sum
+  }
+}
+
+/**
+ * The Abstract base class working as multi-file cloud reading framework
+ * @param conf
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:79]: add description

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+ * @param files PartitionFiles to be read
+ * @param numThreads the number of threads to parallelly read files.
+ * @param maxNumFileProcessed
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:82]: description

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+ * @param filters push down filters
+ * @param execMetrics
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:84]: description

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    for (i <- limit until files.length) {
+      val file = files(i)
+      tasksToRun.enqueue(getBatchRunner(file, conf, filters))
+    }
+    isInitted = true
+    filesToRead = files.length
+  }
+
+  /**
+   * file reading logic in a Callable which will be running in a thread pool
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:129]: upper case File, finish descriptions

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  }
+
+  private def getSizeOfHostBuffers(fileInfo: HostMemoryBuffersWithMetaDataBase): Long = {
+    fileInfo.memBuffersAndSizes.map(_._2).sum
+  }
+
+  private def addNextTaskIfNeeded(): Unit = {
+    if (tasksToRun.size > 0 && !isDone) {
+      val runner = tasksToRun.dequeue()
+      tasks.add(MultiFileThreadPoolFactory.submitToThreadPool(runner, numThreads))
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:217]: this seems odd in abstract code to use the class in parquet file. I think we discussed having a thread pool for each file type so we should abstract this out to call a method that does the right thing based on reader type.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 6]
```diff
@@ -498,36 +498,17 @@ case class GpuParquetPartitionReaderFactory(
...
- * Base classes with common functions for MultiFileParquetPartitionReader and ParquetPartitionReader
- */
-abstract class FileParquetPartitionReaderBase(
-    conf: Configuration,
-    isSchemaCaseSensitive: Boolean,
-    readDataSchema: StructType,
-    debugDumpPrefix: String,
-    execMetrics: Map[String, GpuMetric]) extends PartitionReader[ColumnarBatch] with Logging
-  with ScanWithMetrics with Arm {
+trait FileParquetPartitionReaderBase extends Logging with Arm with ScanWithMetrics {
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:501]: description

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * The base HostMemoryBuffer information read from a single file.
+ */
+trait HostMemoryBuffersWithMetaDataBase {
+  // PartitionedFile to be read
+  def partitionedFile: PartitionedFile
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:45]: why did you change this to partitionedFile vs the 2 individual filePath and partitionValues I had?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 8]
```diff
@@ -1352,27 +1320,22 @@ class MultiFileCloudParquetPartitionReader(
         if (fileBlockMeta.blocks.length == 0) {
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1320]: it does looks like the filePath might not need to be passed back in the ParquetFileInfoWithBlockMeta from filterBlocks here since I think everywhere we use it file is accessible so can just use that.   partValues might be similar so we should check. I think something I missed when adding the other multi-file support.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.connector.read.PartitionReader
+import org.apache.spark.sql.execution.datasources.PartitionedFile
+import org.apache.spark.sql.rapids.InputFileUtils
+import org.apache.spark.sql.rapids.execution.TrampolineUtil
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:40]: Yes, I planned to add both the abstracted classes for coalescing and cloud in this file.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 5 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:217]: Yes, I will fix this issue.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 7 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:45]: Seems the original HostMemoryBuffersWithMetaData has re-defined all of the fields of PartitionedFiles  https://github.com/NVIDIA/spark-rapids/blob/branch-0.5/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala#L1315

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 8 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1320]: we may need them (except partValues) to set input file blocks

``` scala
          InputFileUtils.setInputFileBlock(
            fileBufsAndMeta.partitionedFile.filePath,
            fileBufsAndMeta.partitionedFile.start,
            fileBufsAndMeta.partitionedFile.length)
```

**Comment by tgravescs** [comment]
Note we have started burn down for 0.5 so this should eventually be targeted to 0.6 once that branch is available.  Looks like the sign off check failed.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 8 above)
Comment (tgravescs) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1360]: that is from the HostMemoryBuffersWithMeta right?  I'm talking about ParquetFileInfoWithBlockMeta. I think the filePath in there is only used in places it has direct access to the PartitionedFile, it does look like perhaps the only thing it gets is conversion from String to Path.  Its not a big deal.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 7 above)
Comment (tgravescs) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:46]:
```
case class PartitionedFile(
    partitionValues: InternalRow,
    filePath: String,
    start: Long,
    length: Long,
    @transient locations: Array[String] = Array.empty) {

```

I guess the only thing left now is the locations and it looks like those stay referenced for the entire partition reader so this should be fine then.

Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala:46]: Hi Tom, Could you tell what does "those stay referenced for the entire partition reader " mean? Since PartitionedFile was passed to Task from Driver. so with transient annotation, locations here should be empty. am I right?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 8 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1360]: Oh, Sorry, I misunderstood. I didn't change anything about ParquetFileInfoWithBlockMeta. and I just checked the usage of filePath, and seems it is necessary for Coalescing reading. Let's just keep it.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 4 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala]: > Note we have started burn down for 0.5 so this should eventually be targeted to 0.6 once that branch is available. Looks like the sign off check failed.

Yeah, I will investigate how to abstract coalescing reading which seems more harder than cloud reading.

**Review by wbo4958 — COMMENTED** [review]
> Note we have started burn down for 0.5 so this should eventually be targeted to 0.6 once that branch is available. Looks like the sign off check failed.

Yeah, I will investigate how to abstract coalescing reading which seems more harder than cloud reading.

**Comment by wbo4958** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by GaryShen2008** [comment]
@wbo4958 Can you add more text in the PR to describe more details about why this PR is necessary?

**Comment by wbo4958** [comment]
> @wbo4958 Can you add more text in the PR to describe more details about why this PR is necessary?

Done

**Comment by wbo4958** [comment]
@tgravescs I resolved all the comments, Could you help to review again?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 1 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 3 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 4 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 6 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

**Comment by wbo4958** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 10]
```diff
@@ -1462,131 +1455,56 @@ class MultiFileCloudParquetPartitionReader(
...
-
-    if (memBuffersAndSize.length > 1) {
-      val updatedBuffers = memBuffersAndSize.drop(1)
-      currentFileHostBuffers = Some(fileBufsAndMeta.copy(memBuffersAndSizes = updatedBuffers))
-    } else {
-      currentFileHostBuffers = None
-    }
-    nextBatch
+  /**
+   * file reading logic in a Callable which will be running in a thread pool
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1459]: Not capital luxe File

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 11]
```diff
@@ -1462,131 +1455,56 @@ class MultiFileCloudParquetPartitionReader(
...
+   *
+   * @param file    file to be read
+   * @param conf
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1462]: Add description

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 12]
```diff
@@ -1462,131 +1455,56 @@ class MultiFileCloudParquetPartitionReader(
...
-          } else {
-            batch = readBatch(fileBufsAndMeta)
-            // the data is copied to GPU so submit another task if we were limited
-            addNextTaskIfNeeded()
-          }
+  /**
+   * decode HostMemoryBuffers by GPU
+   *
+   * @param fileBufsAndMeta
+   * @return
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1489]: Update return

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 13]
```diff
@@ -1462,131 +1455,56 @@ class MultiFileCloudParquetPartitionReader(
...
+    ParquetMultiFileThreadPoolFactory.getThreadPool(getLogTag, numThreads)
   }
 
-  private def addNextTaskIfNeeded(): Unit = {
-    if (tasksToRun.nonEmpty && !isDone) {
-      val runner = tasksToRun.dequeue()
-      tasks.add(MultiFileThreadPoolFactory.submitToThreadPool(runner, numThreads))
-    }
-  }
+  override def getLogTag: String = "Parquet"
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:1485]: Done, added some description

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+  protected def fileSystemBytesRead(): Long = {
+    FileSystem.getAllStatistics.asScala.map(_.getThreadStatistics.getBytesRead).sum
+  }
+}
+
+// A tool to create a ThreadPoolExecutor
+// Please note that the TaskContext is not set in these threads and should not be used.
+object MultiFileThreadPoolUtil {
+
+  // create a thread pool
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 10 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 11 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 15]
```diff
@@ -1462,131 +1455,56 @@ class MultiFileCloudParquetPartitionReader(
...
+   */
+  override def getBatchRunner(file: PartitionedFile, conf: Configuration, filters: Array[Filter]):
+      Callable[HostMemoryBuffersWithMetaDataBase] = {
+    new ReadBatchRunner(filterHandler, file, conf, filters)
   }
 
-  private def getSizeOfHostBuffers(fileInfo: HostMemoryBuffersWithMetaData): Long = {
-    fileInfo.memBuffersAndSizes.map(_._2).sum
+  /**
+   * get ThreadPoolExecutor to run the Callable.
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 16]
```diff
@@ -1462,131 +1455,56 @@ class MultiFileCloudParquetPartitionReader(
...
+   *
+   * same ThreadPoolExecutor for cloud and coalescing for same file format
+   * different ThreadPoolExecutors for different file formats
+   *
+   * @return
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala (hunk 12 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

**Comment by wbo4958** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala (hunk 2 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 17]
```diff
@@ -1462,131 +1455,56 @@ class MultiFileCloudParquetPartitionReader(
...
+      Callable[HostMemoryBuffersWithMetaDataBase] = {
+    new ReadBatchRunner(filterHandler, file, conf, filters)
   }
 
-  private def getSizeOfHostBuffers(fileInfo: HostMemoryBuffersWithMetaData): Long = {
-    fileInfo.memBuffersAndSizes.map(_._2).sum
+  /**
+   * get ThreadPoolExecutor to run the Callable.
+   *
+   * same ThreadPoolExecutor for cloud and coalescing for same file format
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

**Comment by wbo4958** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 18]
```diff
@@ -1462,131 +1455,58 @@ class MultiFileCloudParquetPartitionReader(
...
 
-  private def addNextTaskIfNeeded(): Unit = {
-    if (tasksToRun.nonEmpty && !isDone) {
-      val runner = tasksToRun.dequeue()
-      tasks.add(MultiFileThreadPoolFactory.submitToThreadPool(runner, numThreads))
-    }
-  }
+  /**
+   * Get the log prefix
+   *  @return "Parquet"
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 19]
```diff
@@ -1462,131 +1455,58 @@ class MultiFileCloudParquetPartitionReader(
...
   }
 
-  private def addNextTaskIfNeeded(): Unit = {
-    if (tasksToRun.nonEmpty && !isDone) {
-      val runner = tasksToRun.dequeue()
-      tasks.add(MultiFileThreadPoolFactory.submitToThreadPool(runner, numThreads))
-    }
-  }
+  /**
+   * Get the log prefix
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Thx for the info, Done.

**Review by tgravescs — APPROVED** [review]
